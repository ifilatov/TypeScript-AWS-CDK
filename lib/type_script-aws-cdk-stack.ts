import * as cdk from 'aws-cdk-lib/core';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as elasticbeanstalk from 'aws-cdk-lib/aws-elasticbeanstalk';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';
import * as resourcegroups from 'aws-cdk-lib/aws-resourcegroups';
import { Construct } from 'constructs';

export class ThreeTierAppStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a Resource Group
    const resourceGroup = new resourcegroups.CfnGroup(this, 'ResourceGroup', {
      name: 'ThreeTierAppResources',
      resourceQuery: {
        type: 'TAG_FILTERS_1_0',
        query: {
          tagFilters:[
            {
              key:"Application",
              values:["ThreeTierApp"]
            }
          ]
        },
      }
    });

    // Create VPC
    const vpc = new ec2.Vpc(this, 'ThreeTierAppVpc', {
      maxAzs: 2,
      vpcName: 'ThreeTierAppVpc'
    });

    cdk.Tags.of(vpc).add('Application', 'ThreeTierApp');

    // Create Secret Manager for PostgreSQL database credentials
    const dbCreds = new secretsmanager.Secret(this, 'PostgreSQLCreds', {
      secretName: 'PostgreSQLCreds',
      generateSecretString: {
        secretStringTemplate: JSON.stringify({ username: 'postgreAdmin' }),
        excludePunctuation: true,
        includeSpace: false,
        generateStringKey: 'password'
      }
    });

    cdk.Tags.of(dbCreds).add('Application', 'ThreeTierApp');

    // Create RDS PostgreSQL database
    const dbInstance = new rds.DatabaseInstance(this, 'Database', {
      engine: rds.DatabaseInstanceEngine.postgres({ version: rds.PostgresEngineVersion.VER_12 }),
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
      vpc,
      instanceIdentifier: 'three-tier-db',
      credentials: rds.Credentials.fromSecret(dbCreds),
      allocatedStorage: 20, // Adjust as needed
      deletionProtection: false // Set to true for production
    });

    cdk.Tags.of(dbInstance).add('Application', 'ThreeTierApp');

    // Create new role for Elastic Beanstalk
    const ebRole = new iam.Role(this, 'EBRole', {
      assumedBy: new iam.ServicePrincipal('elasticbeanstalk.amazonaws.com'),
    });

    ebRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkWebTier'));
    ebRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkWorkerTier'));
    ebRole.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkMulticontainerDocker'));

    cdk.Tags.of(ebRole).add('Application', 'ThreeTierApp');

    // Add IAM policy statement to grant Elastic Beanstalk access to the dbCreds secret
    ebRole.addToPolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['secretsmanager:GetSecretValue'],
      resources: [dbCreds.secretArn],
    }));

    // Create Instance Profile for Elastic Beanstalk
    const ebInstanceProfile = new iam.CfnInstanceProfile(this, 'CustomInstanceProfile', {
      roles: [ebRole.roleName],
    });

    // Create Elastic Beanstalk application
    const ebsApp = new elasticbeanstalk.CfnApplication(this, 'ThreeTierApp');

    // Create Elastic Beanstalk environment
    const environment = new elasticbeanstalk.CfnEnvironment(this, 'ThreeTierAppEnvironment', {
      applicationName: ebsApp.ref,
      solutionStackName: '64bit Amazon Linux 2023 v6.1.1 running Node.js 20',
      tier: {
        name: 'WebServer',
        type: 'Standard'
      },
      tags: [{
        key:"Application",
        value:"ThreeTierApp"
      }],
      optionSettings: [
        {
            namespace: 'aws:autoscaling:launchconfiguration',
            optionName: 'IamInstanceProfile',
            value: ebInstanceProfile.attrArn,
        },
        {
          namespace: 'aws:elasticbeanstalk:application:environment',
          optionName: 'DB_HOST',
          value: dbInstance.dbInstanceEndpointAddress
        },
        {
          namespace: 'aws:elasticbeanstalk:application:environment',
          optionName: 'DB_PORT',
          value: dbInstance.dbInstanceEndpointPort.toString()
        },
        {
          namespace: 'aws:elasticbeanstalk:application:environment',
          optionName: 'DB_CREDS_SECRET',
          value: dbCreds.secretArn
        }
      ]
    });
  }
}