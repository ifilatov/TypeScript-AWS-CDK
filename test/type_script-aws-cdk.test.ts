import { Template } from 'aws-cdk-lib/assertions';
import * as cdk from 'aws-cdk-lib/core';
import { ThreeTierAppStack } from '../lib/type_script-aws-cdk-stack';

test('Stack Snapshot', () => {
  const app = new cdk.App();
  const stack = new ThreeTierAppStack(app, 'TestStack');
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

test('VPC Created', () => {
  const app = new cdk.App();
  const stack = new ThreeTierAppStack(app, 'TestStack');
  const template = Template.fromStack(stack);
  expect(template.hasResource('AWS::EC2::VPC', {
    "Properties": {
      "CidrBlock":"10.0.0.0/16"
    }
  }));
});

test('Secrets Manager Secret Created', () => {
  const app = new cdk.App();
  const stack = new ThreeTierAppStack(app, 'TestStack');
  const template = Template.fromStack(stack);
  expect(template.hasResource('AWS::SecretsManager::Secret', {
    "Properties": {
      "GenerateSecretString": {
        "ExcludePunctuation":true,
        "GenerateStringKey":"password",
        "IncludeSpace":false
      }
    }
  }));
});

test('RDS Database Instance Created', () => {
  const app = new cdk.App();
  const stack = new ThreeTierAppStack(app, 'TestStack');
  const template = Template.fromStack(stack);
  expect(template.hasResource('AWS::RDS::DBInstance', {
    "Properties": {
      "DBInstanceClass":"db.t3.micro",
      "DBInstanceIdentifier":"three-tier-db",
      "DeletionProtection":false,
      "Engine":"postgres",
      "EngineVersion":"12"
    }
  }));
});

test('Elastic Beanstalk Environment Created', () => {
  const app = new cdk.App();
  const stack = new ThreeTierAppStack(app, 'TestStack');
  const template = Template.fromStack(stack);
  expect(template.hasResource('AWS::ElasticBeanstalk::Environment', {
    "Properties": {
      "ApplicationName": {
        "Ref":"ThreeTierApp"
      },
      "SolutionStackName":"64bit Amazon Linux 2 v3.4.6 running Node.js 14",
      "Tier": {
        "Name":"WebServer",
        "Type":"Standard"
      }
    }
  }));
});