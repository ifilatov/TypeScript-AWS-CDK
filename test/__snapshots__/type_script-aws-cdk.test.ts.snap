// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DatabaseB269D8BB": {
      "DeletionPolicy": "Snapshot",
      "Properties": {
        "AllocatedStorage": "20",
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": "three-tier-db",
        "DBSubnetGroupName": {
          "Ref": "DatabaseSubnetGroup7D60F180",
        },
        "DeletionProtection": false,
        "Engine": "postgres",
        "EngineVersion": "12",
        "MasterUserPassword": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:",
              {
                "Ref": "PostgreSQLCreds12A20BF9",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:",
              {
                "Ref": "PostgreSQLCreds12A20BF9",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "StorageType": "gp2",
        "VPCSecurityGroups": [
          {
            "Fn::GetAtt": [
              "DatabaseSecurityGroup5C91FDCB",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "DatabaseSecurityGroup5C91FDCB": {
      "Properties": {
        "GroupDescription": "Security group for Database database",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DatabaseSubnetGroup7D60F180": {
      "Properties": {
        "DBSubnetGroupDescription": "Subnet group for Database database",
        "SubnetIds": [
          {
            "Ref": "ThreeTierAppVpcPrivateSubnet1SubnetBE8E6545",
          },
          {
            "Ref": "ThreeTierAppVpcPrivateSubnet2Subnet47F69335",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "EBRole71D12AA7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "elasticbeanstalk.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "EBRoleDefaultPolicyA9CC6E70": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "secretsmanager:GetSecretValue",
              "Effect": "Allow",
              "Resource": {
                "Ref": "PostgreSQLCreds12A20BF9",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EBRoleDefaultPolicyA9CC6E70",
        "Roles": [
          {
            "Ref": "EBRole71D12AA7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PostgreSQLCreds12A20BF9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "GenerateSecretString": {
          "ExcludePunctuation": true,
          "GenerateStringKey": "password",
          "IncludeSpace": false,
          "SecretStringTemplate": "{"username":"postgreAdmin"}",
        },
        "Name": "PostgreSQLCreds",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
    "PostgreSQLCredsAttachmentAF8CDDB8": {
      "Properties": {
        "SecretId": {
          "Ref": "PostgreSQLCreds12A20BF9",
        },
        "TargetId": {
          "Ref": "DatabaseB269D8BB",
        },
        "TargetType": "AWS::RDS::DBInstance",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "ThreeTierApp": {
      "Type": "AWS::ElasticBeanstalk::Application",
    },
    "ThreeTierAppEnvironment": {
      "Properties": {
        "ApplicationName": {
          "Ref": "ThreeTierApp",
        },
        "OptionSettings": [
          {
            "Namespace": "aws:elasticbeanstalk:environment",
            "OptionName": "ServiceRole",
            "Value": {
              "Fn::GetAtt": [
                "EBRole71D12AA7",
                "Arn",
              ],
            },
          },
          {
            "Namespace": "aws:elasticbeanstalk:application:environment",
            "OptionName": "DB_HOST",
            "Value": {
              "Fn::GetAtt": [
                "DatabaseB269D8BB",
                "Endpoint.Address",
              ],
            },
          },
          {
            "Namespace": "aws:elasticbeanstalk:application:environment",
            "OptionName": "DB_PORT",
            "Value": {
              "Fn::GetAtt": [
                "DatabaseB269D8BB",
                "Endpoint.Port",
              ],
            },
          },
          {
            "Namespace": "aws:elasticbeanstalk:application:environment",
            "OptionName": "DB_USER",
            "Value": "postgreAdmin",
          },
          {
            "Namespace": "aws:elasticbeanstalk:application:environment",
            "OptionName": "DB_PASSWORD",
            "Value": {
              "Fn::Join": [
                "",
                [
                  "{{resolve:secretsmanager:",
                  {
                    "Ref": "PostgreSQLCreds12A20BF9",
                  },
                  ":SecretString:::}}",
                ],
              ],
            },
          },
        ],
        "PlatformArn": "arn:aws:elasticbeanstalk:us-east-1::platform/Node.js running on 64bit Amazon Linux 2",
        "SolutionStackName": "64bit Amazon Linux 2 v3.4.6 running Node.js 14",
        "Tier": {
          "Name": "WebServer",
          "Type": "Standard",
        },
      },
      "Type": "AWS::ElasticBeanstalk::Environment",
    },
    "ThreeTierAppVpcCE51A4CC": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "ThreeTierAppVpcIGW57CB39ED": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "ThreeTierAppVpcPrivateSubnet1DefaultRoute33051BAC": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "ThreeTierAppVpcPublicSubnet1NATGateway82B7FE87",
        },
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet1RouteTableDAF09179",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "ThreeTierAppVpcPrivateSubnet1RouteTableAssociation447C1755": {
      "Properties": {
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet1RouteTableDAF09179",
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet1SubnetBE8E6545",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "ThreeTierAppVpcPrivateSubnet1RouteTableDAF09179": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "ThreeTierAppVpcPrivateSubnet1SubnetBE8E6545": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "ThreeTierAppVpcPrivateSubnet2DefaultRoute77659965": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "ThreeTierAppVpcPublicSubnet2NATGateway0FE527AD",
        },
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet2RouteTable236FB20B",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "ThreeTierAppVpcPrivateSubnet2RouteTable236FB20B": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "ThreeTierAppVpcPrivateSubnet2RouteTableAssociation2F42E16C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet2RouteTable236FB20B",
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPrivateSubnet2Subnet47F69335",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "ThreeTierAppVpcPrivateSubnet2Subnet47F69335": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "ThreeTierAppVpcPublicSubnet1DefaultRouteB0FA90F8": {
      "DependsOn": [
        "ThreeTierAppVpcVPCGW89ADFF29",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "ThreeTierAppVpcIGW57CB39ED",
        },
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPublicSubnet1RouteTableF0329AA3",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "ThreeTierAppVpcPublicSubnet1EIP33FC0814": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "ThreeTierAppVpcPublicSubnet1NATGateway82B7FE87": {
      "DependsOn": [
        "ThreeTierAppVpcPublicSubnet1DefaultRouteB0FA90F8",
        "ThreeTierAppVpcPublicSubnet1RouteTableAssociation5B2EC895",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "ThreeTierAppVpcPublicSubnet1EIP33FC0814",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPublicSubnet1Subnet0EEE210B",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "ThreeTierAppVpcPublicSubnet1RouteTableAssociation5B2EC895": {
      "Properties": {
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPublicSubnet1RouteTableF0329AA3",
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPublicSubnet1Subnet0EEE210B",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "ThreeTierAppVpcPublicSubnet1RouteTableF0329AA3": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "ThreeTierAppVpcPublicSubnet1Subnet0EEE210B": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "ThreeTierAppVpcPublicSubnet2DefaultRoute76281A05": {
      "DependsOn": [
        "ThreeTierAppVpcVPCGW89ADFF29",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "ThreeTierAppVpcIGW57CB39ED",
        },
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPublicSubnet2RouteTable0C0DA7D2",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "ThreeTierAppVpcPublicSubnet2EIPEC0C41B6": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "ThreeTierAppVpcPublicSubnet2NATGateway0FE527AD": {
      "DependsOn": [
        "ThreeTierAppVpcPublicSubnet2DefaultRoute76281A05",
        "ThreeTierAppVpcPublicSubnet2RouteTableAssociationE89247C7",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "ThreeTierAppVpcPublicSubnet2EIPEC0C41B6",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPublicSubnet2SubnetF3CE3795",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "ThreeTierAppVpcPublicSubnet2RouteTable0C0DA7D2": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "ThreeTierAppVpcPublicSubnet2RouteTableAssociationE89247C7": {
      "Properties": {
        "RouteTableId": {
          "Ref": "ThreeTierAppVpcPublicSubnet2RouteTable0C0DA7D2",
        },
        "SubnetId": {
          "Ref": "ThreeTierAppVpcPublicSubnet2SubnetF3CE3795",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "ThreeTierAppVpcPublicSubnet2SubnetF3CE3795": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            1,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/ThreeTierAppVpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "ThreeTierAppVpcVPCGW89ADFF29": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "ThreeTierAppVpcIGW57CB39ED",
        },
        "VpcId": {
          "Ref": "ThreeTierAppVpcCE51A4CC",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
